src/lib/Helpers.coffee

```diff
@@ -141,40 +124,470 @@ exports.WirePattern = (component, config, proc) ->
       params: 'all'
+
+  config.inPorts = inPorts
+  config.outPorts = outPorts
+
+  # Allow users to selectively fall back to legacy WirePattern implementation
+  if config.legacy or process?.env?.NOFLO_WIREPATTERN_LEGACY
+    platform.deprecated 'noflo.helpers.WirePattern legacy mode is deprecated'
+    setup = legacyWirePattern
+  else
+    setup = processApiWirePattern
+  return setup component, config, proc
+
+# Takes WirePattern configuration of a component and sets up
+# Process API to handle it.
+processApiWirePattern = (component, config, func) ->
+  # Make param ports control ports
+  setupControlPorts component, config
+  # Set up sendDefaults function
+  setupSendDefaults component
+  # Set up bracket forwarding rules
+  setupBracketForwarding component, config
+  component.ordered = config.ordered
+  # Create the processing function
+  component.process (input, output, context) ->
+    # Abort unless WirePattern-style preconditions don't match
+    return unless checkWirePatternPreconditions config, input, output
+    # Populate component.params from control ports
+    component.params = populateParams config, input
+    # Read input data
+    data = getInputData config, input
+    # Read bracket context of first inport
+    groups = getGroupContext component, config.inPorts[0], input
+    if groups.length > 3
+      process.exit 0
+    # Produce proxy object wrapping output in legacy-style port API
+    outProxy = getOutputProxy config.outPorts, output
+
+    debug "WirePattern Process API call with", data, groups, component.params, context.scope
+
+    unless config.async
+      # Set up custom error handlers
+      errorHandler = setupErrorHandler component, config, output
+      # Synchronous WirePattern, call done here
+      func.call component, data, groups, outProxy, null, null, input.scope
+      # No need to call done if component called fail
+      return if output.result.__resolved
+      # Let error handler send any remaining errors
+      do errorHandler
+      # Call done
+      output.done()
+      return
+
+    # Async WirePattern will call the output.done callback itself
+    errorHandler = setupErrorHandler component, config, output
+    func.call component, data, groups, outProxy, (err) ->
+      do errorHandler
+      output.done err
+    , null, null, input.scope
+
+# Provide deprecation warnings on certain more esoteric WirePattern features
+checkDeprecation = (config, func) ->
+  # First check the conditions that force us to fall back on legacy WirePattern
+  if config.group
+    platform.deprecated 'noflo.helpers.WirePattern group option is deprecated. Please port to Process API'
+  if config.field
+    platform.deprecated 'noflo.helpers.WirePattern field option is deprecated. Please port to Process API'
+  # Then add deprecation warnings for other unwanted behaviors
+  if func.length > 4
+    platform.deprecated 'noflo.helpers.WirePattern postpone and resume are deprecated. Please port to Process API'
```

Comment:
Well, except we can't see the argument usage from the outside. Plugged the deprecation warning back in. It will throw anyway if user tries to call the (non-provided) function.